File: tests/lua/input_controller.test.ts
import { PSX_MACHINE_SPEC } from '../../machine/ts/spec/bmsx/model';
import { cartridgeSlots } from '../helpers/cartridge';
import assert from 'node:assert/strict';
import { test } from 'node:test';

import {
	INP_CTRL_ARM,
	INP_CTRL_RESET,
	IO_INP_CTRL,
	IO_INP_KEYS,
	IO_INP_OUTPUT_CTRL,
	IO_INP_OUTPUT_DURATION_MS,
	IO_INP_OUTPUT_INTENSITY_Q16,
	IO_INP_OUTPUT_PORT,
	IO_INP_OUTPUT_STATUS,
	IO_INP_PADS,
	IO_INP_PAD_LX_OFFSET,
	IO_INP_PAD_LY_OFFSET,
	IO_INP_PAD_BUTTONS_OFFSET,
	IO_INP_POINTER_BUTTONS,
	IO_INP_POINTER_WHEEL,
	IO_INP_POINTER_X,
	IO_INP_POINTER_Y,
	IO_INP_STATUS,
} from '../../machine/ts/spec/bmsx/io';
import { IO_WORD_SIZE } from '../../machine/ts/spec/bmsx/memory_map';
import { encodeSignedFix16 } from '../../machine/ts/machine/common/numeric';
import { AcceptedInterruptKind } from '../../machine/ts/machine/cpu/cpu';
import { InputController } from '../../machine/ts/machine/devices/input/controller';
import {
	INP_OUTPUT_CTRL_APPLY,
	INP_POINTER_BUTTON_PRIMARY,
	INP_POINTER_BUTTON_SECONDARY,
	INPUT_CONTROLLER_KEY_WORD_COUNT,
	INPUT_CONTROLLER_OUTPUT_INTENSITY_Q16_ONE,
	InputControllerGamepadAxis,
	type InputControllerInputSource,
	InputControllerSampleContext,
	type InputControllerSnapshot,
} from '../../machine/ts/machine/devices/input/contracts';
import { Memory } from '../../machine/ts/machine/memory/memory';
import { Machine } from '../../machine/ts/machine/machine';

const HID_KEY_X = 27;
const HID_KEY_F2 = 59;
const PAD_A_BIT = 0;

type FakeVibration = {
	padIndex: number;
	durationMs: number;
	intensity: number;
};

function writeSample(snapshot: InputControllerSnapshot, keyWords: Uint32Array): void {
	snapshot.keyWords.set(keyWords);
	snapshot.pointerButtons = (1 << INP_POINTER_BUTTON_PRIMARY) | (1 << INP_POINTER_BUTTON_SECONDARY);
	snapshot.pointerXQ16 = encodeSignedFix16(12.5);
	snapshot.pointerYQ16 = encodeSignedFix16(-3.25);
	snapshot.pointerWheelQ16 = encodeSignedFix16(1.5);
	snapshot.rumbleSupportMask = 1 << 2;
	snapshot.pads[0].buttons = 1 << PAD_A_BIT;
	snapshot.pads[0].axesQ16[InputControllerGamepadAxis.LeftX] =
		encodeSignedFix16(-0.5);
	snapshot.pads[0].axesQ16[InputControllerGamepadAxis.LeftY] =
		encodeSignedFix16(0.25);
}

function createHarness(): {
	memory: Memory;
	machine: Machine;
	controller: InputController;
	vibrations: FakeVibration[];
	samples: () => number;
	sampleContexts: () => readonly InputControllerSampleContext[];
	setKey: (usage: number, down: boolean) => void;
	setSupervisorRequestLine: (high: boolean) => void;
} {
	const memory = new Memory({ systemRom: new Uint8Array(0), cartridgeSlots: cartridgeSlots() }, PSX_MACHINE_SPEC.ramBytes);
	const vibrations: FakeVibration[] = [];
	const keyWords = new Uint32Array(INPUT_CONTROLLER_KEY_WORD_COUNT);
	keyWords[HID_KEY_X >>> 5] = 1 << (HID_KEY_X & 31);
	let sampleCount = 0;
	const sampleContexts: InputControllerSampleContext[] = [];
	let supervisorRequestLineHigh = false;
	const input: InputControllerInputSource = {
		sampleInputControllerSnapshot(
			snapshot: InputControllerSnapshot,
			context: InputControllerSampleContext,
		) {
			sampleCount += 1;
			sampleContexts.push(context);
			writeSample(snapshot, keyWords);
		},
		supervisorRequestLineHigh() {
			return supervisorRequestLineHigh;
		},
		applyInputControllerVibrationEffect(padIndex, durationMs, intensity) {
			vibrations.push({ padIndex, durationMs, intensity });
		},
	};
	const machine = new Machine(memory, input, PSX_MACHINE_SPEC);
	machine.resetDevices();
	const controller = machine.inputController;
	return {
		memory,
		machine,
		controller,
		vibrations,
		samples: () => sampleCount,
		sampleContexts: () => sampleContexts,
		setKey: (usage, down) => {
			const mask = 1 << (usage & 31);
			const word = usage >>> 5;
			keyWords[word] = down ? keyWords[word] | mask : keyWords[word] & ~mask;
		},
		setSupervisorRequestLine: high => {
			supervisorRequestLineHigh = high;
		},
	};
}

test('input controller latches one raw MMIO snapshot on armed VBlank', () => {
	const live = createHarness();
	live.setKey(HID_KEY_F2, true);
	live.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_ARM);
	live.controller.onVblankEdge(77);

	assert.equal(live.samples(), 1);
	assert.equal(live.memory.readIoU32(IO_INP_STATUS), 1);
	assert.equal(live.memory.readIoU32(IO_INP_KEYS + (HID_KEY_X >>> 5) * IO_WORD_SIZE), 1 << (HID_KEY_X & 31));
	assert.notEqual(live.memory.readIoU32(IO_INP_KEYS + (HID_KEY_F2 >>> 5) * IO_WORD_SIZE) & (1 << (HID_KEY_F2 & 31)), 0);
	assert.equal(live.memory.readIoU32(IO_INP_POINTER_BUTTONS), (1 << INP_POINTER_BUTTON_PRIMARY) | (1 << INP_POINTER_BUTTON_SECONDARY));
	assert.equal(live.memory.readIoU32(IO_INP_POINTER_X), encodeSignedFix16(12.5));
	assert.equal(live.memory.readIoU32(IO_INP_POINTER_Y), encodeSignedFix16(-3.25));
	assert.equal(live.memory.readIoU32(IO_INP_POINTER_WHEEL), encodeSignedFix16(1.5));
	assert.equal(live.memory.readIoU32(IO_INP_PADS + IO_INP_PAD_BUTTONS_OFFSET), 1 << PAD_A_BIT);
	assert.equal(live.memory.readIoU32(IO_INP_PADS + IO_INP_PAD_LX_OFFSET), encodeSignedFix16(-0.5));
	assert.equal(live.memory.readIoU32(IO_INP_PADS + IO_INP_PAD_LY_OFFSET), encodeSignedFix16(0.25));
	assert.equal(live.memory.readIoU32(IO_INP_OUTPUT_STATUS), 1 << 2);
});

test('input controller save-state restores raw latch registers', () => {
	const live = createHarness();
	live.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_ARM);
	live.controller.onVblankEdge(77);
	live.memory.writeMappedWord(IO_INP_OUTPUT_PORT, 2);
	live.memory.writeMappedWord(IO_INP_OUTPUT_INTENSITY_Q16, INPUT_CONTROLLER_OUTPUT_INTENSITY_Q16_ONE >>> 1);
	live.memory.writeMappedWord(IO_INP_OUTPUT_DURATION_MS, 120);

	const savedInput = live.controller.captureState();
	const restored = createHarness();
	restored.controller.restoreState(savedInput);

	assert.equal(restored.memory.readIoU32(IO_INP_STATUS), 1);
	assert.equal(restored.memory.readIoU32(IO_INP_KEYS + (HID_KEY_X >>> 5) * IO_WORD_SIZE), 1 << (HID_KEY_X & 31));
	assert.equal(restored.memory.readIoU32(IO_INP_POINTER_X), encodeSignedFix16(12.5));
	assert.equal(restored.memory.readIoU32(IO_INP_PADS), 1 << PAD_A_BIT);
	assert.equal(restored.memory.readIoU32(IO_INP_OUTPUT_PORT), 2);
	assert.equal(restored.memory.readIoU32(IO_INP_OUTPUT_INTENSITY_Q16), INPUT_CONTROLLER_OUTPUT_INTENSITY_Q16_ONE >>> 1);
	assert.equal(restored.memory.readIoU32(IO_INP_OUTPUT_DURATION_MS), 120);
});

test('input controller output registers emit selected-pad vibration commands', () => {
	const live = createHarness();
	live.memory.writeMappedWord(IO_INP_OUTPUT_PORT, 2);
	live.memory.writeMappedWord(IO_INP_OUTPUT_INTENSITY_Q16, INPUT_CONTROLLER_OUTPUT_INTENSITY_Q16_ONE >>> 1);
	live.memory.writeMappedWord(IO_INP_OUTPUT_DURATION_MS, 120);

	live.memory.writeMappedWord(IO_INP_OUTPUT_CTRL, INP_OUTPUT_CTRL_APPLY);
	assert.deepEqual(live.vibrations, [{ padIndex: 2, durationMs: 120, intensity: 0.5 }]);
	assert.equal(live.memory.readIoU32(IO_INP_OUTPUT_CTRL), 0);
});

test('input controller exposes the VBlank sample edge without leaking the sample latch', () => {
	const harness = createHarness();
	harness.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_ARM);
	assert.equal(harness.controller.captureState().sampleArmed, true);
	harness.controller.cancelSampleArm();
	assert.equal(harness.controller.captureState().sampleArmed, false);
});

test('input controller raises one supervisor request edge and the device fence vectors NMI', () => {
	const harness = createHarness();
	const noInterrupt = AcceptedInterruptKind.None;
	const restoredState = harness.controller.captureState();
	restoredState.supervisorRequestLineHigh = true;
	harness.controller.restoreState(restoredState);
	harness.setSupervisorRequestLine(true);
	harness.controller.onVblankEdge(1);

	assert.equal(harness.samples(), 0);
	assert.equal(harness.memory.readIoU32(IO_INP_STATUS), 0);
	assert.equal(harness.memory.readIoU32(IO_INP_KEYS + (HID_KEY_F2 >>> 5) * IO_WORD_SIZE), 0);
	assert.equal(harness.machine.cpu.peekPendingInterrupt(), noInterrupt);
	harness.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_RESET);
	harness.controller.onVblankEdge(2);
	assert.equal(harness.machine.cpu.peekPendingInterrupt(), noInterrupt);

	harness.setSupervisorRequestLine(false);
	harness.controller.onVblankEdge(3);
	harness.setSupervisorRequestLine(true);
	harness.controller.onVblankEdge(4);

	assert.equal(harness.machine.cpu.peekPendingInterrupt(), noInterrupt);
	harness.machine.systemController.onService();
	assert.equal(harness.machine.cpu.peekPendingInterrupt(), AcceptedInterruptKind.NonMaskable);
	assert.equal(harness.controller.captureState().supervisorRequestLineHigh, true);
});

test('input controller derives source-sampling context from the supervisor phase', () => {
	const harness = createHarness();
	harness.setSupervisorRequestLine(true);
	harness.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_ARM);
	harness.controller.onVblankEdge(1);
	assert.deepEqual(harness.sampleContexts(), [InputControllerSampleContext.Normal]);
	assert.equal(harness.machine.systemController.supervisorContextActive(), true);

	harness.memory.writeMappedWord(IO_INP_CTRL, INP_CTRL_ARM);
	harness.controller.onVblankEdge(2);
	assert.deepEqual(harness.sampleContexts(), [
		InputControllerSampleContext.Normal,
		InputControllerSampleContext.Supervisor,
	]);
});

File: tests/cpp/input_controller_test.cpp
#include "spec/bmsx/io.h"
#include "machine/machine.h"
#include "spec/bmsx/memory_map.h"
#include "machine/memory/memory.h"
#include "spec/bmsx/model.h"
#include "support/cartridge_fixture.h"

#include <array>
#include <stdexcept>

namespace {

void require(bool condition, const char* message) {
	if (!condition) {
		throw std::runtime_error(message);
	}
}

class TestInput final : public bmsx::InputControllerInputSource {
public:
	void sampleInputControllerSnapshot(
		bmsx::InputControllerSnapshot& snapshot,
		bmsx::InputControllerSampleContext context) override {
		sampleContexts[fullSampleCount] = context;
		fullSampleCount += 1;
		snapshot.keyWords = keyWords;
		snapshot.pointerXQ16 = 0x000c8000u;
		snapshot.pointerYQ16 = 0xfffcc000u;
		snapshot.pads[0].axesQ16[static_cast<size_t>(
			bmsx::InputControllerGamepadAxis::LeftX)] = 0xffff8000u;
	}

	auto supervisorRequestLineHigh() const -> bool override {
		return supervisorRequestLine;
	}

	void applyInputControllerVibrationEffect(bmsx::i32, bmsx::f64, bmsx::f32) override {
	}

	void setKey(bmsx::u32 usage, bool down) {
		const size_t word = usage >> 5u;
		const bmsx::u32 mask = 1u << (usage & 31u);
		if (down) {
			keyWords[word] |= mask;
		} else {
			keyWords[word] &= ~mask;
		}
	}

	std::array<bmsx::u32, bmsx::INPUT_CONTROLLER_KEY_WORD_COUNT> keyWords{};
	std::array<bmsx::InputControllerSampleContext, 2> sampleContexts{};
	int fullSampleCount = 0;
	bool supervisorRequestLine = false;
};

struct InputControllerHarness {
	std::array<bmsx::u8, 1> emptyRom{{0}};
	bmsx::Memory memory;
	TestInput input;
	bmsx::Machine machine;

	InputControllerHarness()
		: memory(
			bmsx::MemoryInit{ { emptyRom.data(), 0u }, bmsx::test::cartridgeSlots() },
			bmsx::PSX_MACHINE_SPEC.ramBytes)
		, machine(memory, input, bmsx::PSX_MACHINE_SPEC) {
		machine.resetDevices();
	}
};

void testSystemNmiEdgeDoesNotPublishAnUnarmedSnapshot() {
	InputControllerHarness harness;
	TestInput& input = harness.input;
	bmsx::CPU& cpu = harness.machine.cpu;
	bmsx::InputController& controller = harness.machine.inputController;
	constexpr bmsx::AcceptedInterruptKind noInterrupt =
		bmsx::AcceptedInterruptKind::None;
	bmsx::InputControllerState restored = controller.captureState();
	restored.supervisorRequestLineHigh = true;
	controller.restoreState(restored);
	input.supervisorRequestLine = true;
	controller.onVblankEdge(1u);

	require(input.fullSampleCount == 0, "unarmed VBlank does not sample the full input frame");
	require(harness.memory.readIoU32(bmsx::IO_INP_STATUS) == 0u, "unarmed VBlank leaves the sample sequence unchanged");
	require(cpu.peekPendingInterrupt() == noInterrupt, "a restored high request line is not a new edge");

	harness.memory.writeMappedU32LE(bmsx::IO_INP_CTRL, bmsx::INP_CTRL_RESET);
	controller.onVblankEdge(2u);
	require(cpu.peekPendingInterrupt() == noInterrupt, "guest ICU reset cannot synthesize a physical edge");

	input.supervisorRequestLine = false;
	controller.onVblankEdge(3u);
	input.supervisorRequestLine = true;
	controller.onVblankEdge(4u);
	require(cpu.peekPendingInterrupt() == noInterrupt, "the ICU edge waits for the common device fence");
	harness.machine.systemController.onService();
	require(cpu.peekPendingInterrupt() == bmsx::AcceptedInterruptKind::NonMaskable, "the completed fence requests NMI");
	require(controller.captureState().supervisorRequestLineHigh, "save-state retains the physical edge level");
}

void testArmedVblankPublishesTheFullSnapshot() {
	constexpr bmsx::u32 f2Usage = 59u;
	InputControllerHarness harness;
	harness.input.setKey(f2Usage, true);
	harness.memory.writeMappedU32LE(bmsx::IO_INP_CTRL, bmsx::INP_CTRL_ARM);
	harness.machine.inputController.onVblankEdge(7u);

	require(harness.input.fullSampleCount == 1, "armed VBlank samples one full input frame");
	require(
		harness.input.sampleContexts[0] == bmsx::InputControllerSampleContext::Normal,
		"initial BIOS execution uses the normal source-sampling context"
	);
	require(harness.memory.readIoU32(bmsx::IO_INP_STATUS) == 1u, "armed VBlank advances the sample sequence");
	const bmsx::u32 f2WordAddress = bmsx::IO_INP_KEYS + (f2Usage >> 5u) * bmsx::IO_WORD_SIZE;
	require((harness.memory.readIoU32(f2WordAddress) & (1u << (f2Usage & 31u))) != 0u,
		"armed VBlank publishes F2 in the raw guest bitmap");
	require(harness.memory.readIoU32(bmsx::IO_INP_POINTER_X) == 0x000c8000u,
		"armed VBlank latches the source-port pointer word directly");
	require(harness.memory.readIoU32(bmsx::IO_INP_POINTER_Y) == 0xfffcc000u,
		"armed VBlank preserves signed source-port pointer bits");
	require(harness.memory.readIoU32(bmsx::IO_INP_PADS + bmsx::IO_INP_PAD_LX_OFFSET) == 0xffff8000u,
		"armed VBlank latches the source-port axis word directly");
}

void testSupervisorPhaseSelectsTheSourceSamplingContext() {
	InputControllerHarness harness;
	harness.input.supervisorRequestLine = true;
	harness.memory.writeMappedU32LE(bmsx::IO_INP_CTRL, bmsx::INP_CTRL_ARM);
	harness.machine.inputController.onVblankEdge(1u);
	require(
		harness.input.sampleContexts[0] == bmsx::InputControllerSampleContext::Normal,
		"the request edge samples before entering the supervisor transition"
	);
	require(
		harness.machine.systemController.supervisorContextActive(),
		"the physical edge starts the retained supervisor transition"
	);

	harness.memory.writeMappedU32LE(bmsx::IO_INP_CTRL, bmsx::INP_CTRL_ARM);
	harness.machine.inputController.onVblankEdge(2u);
	require(
		harness.input.sampleContexts[1] == bmsx::InputControllerSampleContext::Supervisor,
		"the active transition selects the supervisor source-sampling context"
	);
}

} // namespace

int main() {
	testSystemNmiEdgeDoesNotPublishAnUnarmedSnapshot();
	testArmedVblankPublishesTheFullSnapshot();
	testSupervisorPhaseSelectsTheSourceSamplingContext();
	return 0;
}
